### DISCUSSION TITLE
Duplex voice calls in Sideband

### DISCUSSION THREAD

**Discussion by Skretch1974** [markqvist/Reticulum#563] [Requests for New Features]
Hi. There is a proposal to add to the Sideband application the ability to make duplex voice calls with the same list of codecs and bandwidth control by their speed, as it is now implemented in the 
Reticulum-meshchat. Our team is actively using and testing voice calls and it works just fine even with the use of mobile internet on a 2G (edge) connection, which currently has the best coverage in remote areas away from towns and cities. It is also now possible to bridge WiFi radio bridges over long distances and create separate isolated Lan networks through which voice calls can be made. 
I also suggest to consider adding another transport interface using a regular telephone audio channel as a transport, as implemented in the jackpair project - https://github.com/gegel/jackpair.
Here through any analog audio channel a digital duplex channel is created, through which all communication on low-bitrate codecs for voice calls or data and text messages is carried out.

**Comment by resiliencetheatre** [comment]
I really love to participate this without using WebRTC. Maybe I could check how current implementation pipes packets between entities.

**Comment by markqvist** [comment]
For a while, I've been working on a protocol tentatively dubbed "RMV", Reticulum Voice Mesh. You could think of this as the real-time audio equivalent of LXMF. The reason that I have not yet implemented real-time audio in Sideband is, that when I eventually do, I want it to be based on a flexible, open and extensible protocol, like the messaging is currently with LXMF.

Some of the primary design for RVM are:

- Work natively over Reticulum
- Provide a secure, universal and flexible Reticulum-native protocol for real-time audio
- Provide both simplex and duplex call functionality
- Provide group/conference call functionality
- Provide functionality for broadcast audio (think public radio broadcasts over ISM and license-free spectrum)
- Provide a variety of audio encodings based on open and free standards
- Be easy to integrate with legacy VoIP/POTS phone systems

I have no idea how far out an actual release of this is though.

**Comment by Skretch1974** [comment]
Hello. There are some peculiarities in realization of voice transmission in low bandwidth networks. A few years ago one author under the pseudonym Gegel independently implemented a unique at that time project on voice transmission through TOR networks, having developed his own modernized codec based on Melpe codec, able to work in networks with delays more than 3 seconds, and also developed his own protocol over tcp for voice transmission. You can read more about his research with the help of an interpreter at https://habr.com/ru/articles/448856/.
He has also implemented a software audio modem to establish a digital duplex communication channel over a standard telephone audio channel, for example, one can make an audio call on Skype and run a digital data channel for both voice and text messages over this connection. This is potentially another type of transport layer for the reticulum project, allowing, for example, from an area where there is no internet or digital radio modem coverage but a standard telephone wireline connection, to implement a digital gateway to the reticulum. 
If you, as a developer, are interested to see Mr. Hegel's developments, I have kept a complete archive of his files and developments. https://web.archive.org/web/20160113114629/http://torfone.org

**Reply by markqvist** [reply]
That is interesting work. Thanks for the references, and for the archival effort! I will look into it.

**Comment by yuz86** [comment]
Hi. Could you please advise me how I can activate voice calls in the new version of Sideband? I go to the program settings, but the switch that activates voice calls is grayed out and can not be turned on, it is not active, respectively, in all other sections of the program also nothing is displayed. 
Regards Jon.
![2](https://github.com/user-attachments/assets/3b3a5883-24ae-4623-86bc-6ce27fc42738)
![3](https://github.com/user-attachments/assets/c2b53fde-81e9-4848-8f41-4640f4d92e2c)
![1](https://github.com/user-attachments/assets/053acb9d-b75f-4129-b208-8f5836f440a9)

**Reply by markqvist** [reply]
LXST, the realtime voice protocol powering the voice call functionality uses a new audio backend I wrote specifically for it. That backend is not implemented fully on Android yet. When it is, the voice call functionality will be available there as well.

**Comment by yuz86** [comment]
Thank you for the clarification, I would also like to suggest that as an optimization and convenience of working with text messages in Sideband to consider the possibility of hiding service information (Sent, State, Audio Message) from the main chat window and hide it under three dots to the right of each message, and the delivery status to do as on all messengers in the form of two underlining small ticks in the lower right part of the message. This way the information field will not be overloaded with unnecessary information and if necessary it can be easily viewed by calling the corresponding menu under the three dots. 
Thank you for a very necessary and timely application, which may be very useful in case of emergency events. Regards to you.

**Reply by markqvist** [reply]
Just a small tip, if things are a bit too crowded on the screen, you can adjust the UI scaling in preferences. This will universally scale the entire interface, which can help on screens where the general system scaling is set to display everything in rather large fonts, etc.

**Comment by yuz86** [comment]
Thank you. I also wanted to ask if you consider the possibility of realizing in your project the possibility of creating a digital secure communication channel through audio channel, for example, during a regular GSM call, as it is realized in the jackpair project. 
https://www.jackpair.com/
Or the alternative project above in this topic by Mr. Gegel 
https://github.com/gegel/jackpair
Nowadays there are more and more cases when the authorities of many countries disconnect access to the Internet, both mobile and home providers, but the usual voice communication still works, this analog channel can be used by connecting a low-speed modem to it and send Sideband traffic through it. And the call for example will be made to another city to the working gateway, where there is access to the Internet, thus the delivery of messages will be carried out.
[vhf_crypt_win.zip](https://github.com/user-attachments/files/20233094/vhf_crypt_win.zip)
[jackpair_win.zip](https://github.com/user-attachments/files/20233095/jackpair_win.zip)

**Reply by worst-git-server** [reply]
That's an interesting idea. For people who still have old computers with an RJ11 cable, it could be possible to make a ModemInterface. Reticulum works very well on slow communication methods so that would be possible and would be a good way to communicate and stay under the radar.

**Comment by yuz86** [comment]
Hi. I tested today both of the utilities attached in my previous post that I downloaded from Mr. Gegel's website and they really work, I tested between my two computers by calling via Telegram and connected the audio devices of the microphone and the phone via a virtual sound card, surprisingly the transmitted sound is quite good, even in duplex mode! I really hope that my idea of implementing such an interface in your projects will be supported by the community, after all it is a very reliable additional communication channel for Retikulum applications.

**Reply by helmick43270** [reply]
Not all services offering voice allow digital signals to pass. Would be funny if someone came up with stegano-digital voice mode. Would sound like two people talking nonsense, but would decode to actual conversation at endpoints!

**Reply by worst-git-server** [reply]
> Not all services offering voice allow digital signals to pass. Would be funny if someone came up with stegano-digital voice mode. Would sound like two people talking nonsense, but would decode to actual conversation at endpoints!

Here's a way I can think of, all it requires is 2 shared sound files(A and B), and some encryption key(s). It would work as follows:
1. Person A will stream the sound file A but will also add some additional noise, the actual (encrypted) conversation.
2. Person B will, using sound file A extract the noise and decrypt it(then play it on the speakers).

The exact same thing will be done with person B(and file B).
The sound files could be created by recording each side of a not-so-secret conversation.
